module/zfs/arc.c

```diff
@@ -7249,9 +7254,45 @@ arc_state_multilist_index_func(multilist_t *ml, void *obj)
 static void
 arc_tuning_update(void)
 {
-	uint64_t allmem = arc_all_memory();
+	uint64_t new_allmem = arc_all_memory();
 	unsigned long limit;
 
+	if (!allmem)
+		allmem = new_allmem;
+
+	if ((zfs_dirty_data_max_max) &&
+	    (zfs_dirty_data_max_max != dirty_data_max_max))
+		dirty_data_max_max = zfs_dirty_data_max_max;
+
+	if ((zfs_dirty_data_max) &&
+	    (zfs_dirty_data_max != dirty_data_max))
+		dirty_data_max = zfs_dirty_data_max;
+
+	/*
+	 * Memory hotplug. The only case to consider is
+	 * when memory is increased as downgrading memory
+	 * is anyways disruptive
+	 */
+	if (new_allmem > allmem) {
+		/*
+		 * If there is a custom value set, we shouldn't be
+		 * touching it. Whoever set it, should be responsible
+		 * for setting it to scale with the memory.
+		 */
+		allmem = new_allmem;
+
+		if (zfs_dirty_data_max_max == 0)
+			dirty_data_max_max = MIN(ZFS_DIRTY_MAX_LIMIT,
+			    allmem * zfs_dirty_data_max_max_percent / 100);
+
+		if (zfs_dirty_data_max == 0) {
+			dirty_data_max = allmem *
+			    zfs_dirty_data_max_percent / 100;
+			dirty_data_max = MIN(dirty_data_max,
+			    dirty_data_max_max);
+		}
+	}
```

Comment:
What do you think about moving this dirty tuning logic in to a `dsl_pool_tuning()_update()` function in `dsl_pool.c`.